### PR TITLE
8242220: Capture JFR stack trace and thread other than the one the event is committed on

### PR DESCRIPTION
**Pull request by jbachorik**
This is an early prototype of an enhanced Event.commit(thread) method which allows committing a JFR event on a custom java thread. This would allow, for example, having a scheduled 'sampling' thread generating generating events and filling them up with shared aggregation data and then committing them on behalf of a 'sampled' thread.

Committing an event on the given thread means that the event will be associated with that particular thread, will receive the stack trace of that thread at that particular moment and, internally, will be written out using that thread's thread-local structures.

The change is rather trivial - extend the Event API to provide commit(Thread) method and then use thread-local handshake to fill in the stacktrace details of the indicated thread. The code changes are, however, more extensive as all the supporting classes need to be also aware of the new Event method.

As already mentioned this is a very early prototype of this particular functionality. I am aware that finalizing this will take much more work but at least for DataDog this feature is quite useful and starting the discussion about it with a concrete piece of code seems to be the right way to go.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ❌ (6/6 failed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (6/6 passed) | ✔️ (3/3 passed) |

**Failed test tasks**
- [Linux x64 (build debug)](https://github.com/jbachorik/jdk/runs/1459000343)
- [Linux x64 (build hotspot minimal)](https://github.com/jbachorik/jdk/runs/1459000446)
- [Linux x64 (build hotspot no-pch)](https://github.com/jbachorik/jdk/runs/1459000374)
- [Linux x64 (build hotspot optimized)](https://github.com/jbachorik/jdk/runs/1459000477)
- [Linux x64 (build hotspot zero)](https://github.com/jbachorik/jdk/runs/1459000399)
- [Linux x64 (build release)](https://github.com/jbachorik/jdk/runs/1459000317)

### Issue
 * [JDK-8242220](https://bugs.openjdk.java.net/browse/JDK-8242220): Capture JFR stack trace and thread other than the one the event is committed on

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/741/head:pull/741`
`$ git checkout pull/741`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -14,3 +14,8 @@ test/nashorn/lib
 NashornProfile.txt
 **/JTreport/**
 **/JTwork/**
+*.class
+# Eclipse specific folders
+.classpath
+.settings
+.project
```
Comment (magicus) [inline .gitignore:21]: Just a note: changes to .gitignore does not belong in this PR. If you really intend to get this changes into the JDK (and not just as part of the Draft PR for development purposes), please file a separate issue for this.

Comment (jbachorik) [reply .gitignore:22]: This will be definitely removed for the final PR. I would leave it here for a while because it prevents a bunch of junk being reported by `git status`.
